### PR TITLE
RavenDB-16274 - reorganized smuggler inheritance and properly support import

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-16274

### Additional description

Sharded import will open stream against each of the shards and write the data to the proper shard directly. 

We aren't addressing here the issue of keeping the TCP connection alive for shards that didn't receive anything for long period of time, since we plan to rely on the framework to do that for us https://issues.hibernatingrhinos.com/issue/RavenDB-18088

### Type of change

- Optimization

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- existing tests should cover this issue
- 
### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

**Review by efratshenhar — APPROVED** [review]

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 1]
```diff
@@ -130,9 +130,7 @@ public ICounterActions Counters(SmugglerResult result)
 
         public ICounterActions LegacyCounters(SmugglerResult result)
         {
-            // Used only in Stream Destination, needed when we writing from Stream Source to Stream Destination
-            throw new NotSupportedException("LegacyCounters is not supported in Database destination, " +
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:134]: can you explain why this is OK now?

File: src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs [hunk 2]
```diff
@@ -38,19 +32,11 @@ public class ShardedDatabaseSmuggler : SmugglerBase
         private readonly List<DatabaseRecord.ShardRangeAssignment> _shardAllocation;
         private readonly DatabaseRecord _databaseRecord;
         private readonly ServerStore _server;
-        private int _itemsInBatch;
         private long _buildVersion;
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs]: this is not used, will this not be a problem? or maybe we are handling that in base?

File: src/Raven.Server/Smuggler/Migration/Migrator_V2.cs [hunk 3]
```diff
@@ -100,7 +100,7 @@ private async Task MigrateAttachments(string lastEtag, SmugglerResult parameters
                 SkipRevisionCreation = true
             };
 
-            destination.InitializeAsync(options, parametersResult, buildVersion: default);
+            await destination.InitializeAsync(options, parametersResult, buildVersion: default);
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Migration/Migrator_V2.cs]: don't we need to put this in using?

File: src/Raven.Server/Smuggler/Migration/Migrator_V3.cs [hunk 4]
```diff
@@ -122,7 +122,7 @@ private async Task<string> MigrateRavenFs(string lastEtag, SmugglerResult parame
                 SkipRevisionCreation = true
             };
 
-            destination.InitializeAsync(options, parametersResult, _buildVersion);
+            await destination.InitializeAsync(options, parametersResult, _buildVersion);
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Migration/Migrator_V3.cs]: don't we need to put this in using?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:134]: previously the only valid destination for `LegacyCounters` could be only the database, where we could convert that on the fly from `LegacyCounters` to `GroupCounters`.

Now we support stream->stream smuggler, and we can't do the conversion on the fly.

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 5]
```diff
@@ -1844,6 +1844,53 @@ public virtual void Dispose()
             _disposeOnceRunner.Dispose();
         }
 
+        public delegate HttpMessageHandler CreateMessageHandler(X509Certificate2 certificate, bool setSslProtocols, bool useCompression,
+            bool hasExplicitlySetCompressionUsage = false);
+
+        public CreateMessageHandler CreateMessageHandlerDelegate = CreateHttpMessageHandler;
+        /*
+        public static SocketsHttpHandler CreateSocketMessageHandler(X509Certificate2 certificate, bool setSslProtocols, bool useCompression, bool hasExplicitlySetCompressionUsage = false)
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:1852]: Why the commented code?

File: src/Raven.Server/Documents/Sharding/ShardExecutor.cs [hunk 6]
```diff
@@ -18,6 +18,21 @@ public ShardExecutor(ShardedContext shardedContext)
...
+            try
+            {
+                await executor.ExecuteAsync(cmd, ctx);
+                return cmd.Result;
+            }
+            finally
+            {
+                release.Dispose();
+            }
+        }
```
Comment (ayende) [inline src/Raven.Server/Documents/Sharding/ShardExecutor.cs:34]: using?

File: src/Raven.Server/Smuggler/Documents/MultiShardedDestination.cs [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+        {
+            var stream = _handler.GetOutputStream(holders[shard].OutStream.OutputStream.Result, _options);
+            holders[shard].InputStream = stream;
+            holders[shard].ContextReturn = _handler.ContextPool.AllocateOperationContext(out JsonOperationContext context);
+            var destination = new StreamDestination(stream, context, _source);
+            holders[shard].DestinationAsyncDisposable = await destination.InitializeAsync(_options, result, buildVersion);
+            _destinations[shard] = destination;
+        }
+
+        internal struct StreamDestinationHolder : IAsyncDisposable
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/MultiShardedDestination.cs:73]: Why is that a struct? If this is `IAsyncDisposable`, it is going to be boxed for sure.

**Pull request merged**